### DISCUSSION TITLE
unlimited argument for an option with comma spliter

### DISCUSSION THREAD

**Discussion by Alirezaaraby** [fastapi/typer#1341] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the Typer documentation, with the integrated search.
- [X] I already searched in Google "How to X in Typer" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to Typer but to [Click](https://github.com/pallets/click).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
import typer

def main(command: List[str] = typer.Option(None)):
    print(command)
```

### Description

I want to pass unlimited multiple arguments to `type.option` with comma splitter like below.

```shell
$ python cli.py add test -C "hi I'm test", "hello this is test"
```

I want to get a list like `["hi I'm test","hello this is test"]` from the above input.

### Operating System

Windows

### Operating System Details

Windows Version 10.0.19045.2486

### Typer Version

0.7.0

### Python Version

Python 3.9.1

**Comment by heiskane** [comment]
As far as i understand the intended way to use a list type parameter would be like this:
```bash
$ python app.py --command arg1 --command arg2
['arg1', 'arg2']
```

If you really want to use comma separated strings you could do the following:
```python
from typing import Annotated

import typer

def comma_list(raw: str) -> list[str]:
    return raw.split(",")


def main(
    command: Annotated[list, typer.Option(parser=comma_list)],
):
    print(command)


if __name__ == "__main__":
    typer.run(main)
```

```bash
$ python potato/main.py --command "hello","world"
['hello', 'world']
```

Note that using `list[str]` instead of `list` for the `command` parameter seems to attempt some type conversion magic resulting in a nested list so this isn't perfect. Personally I would just keep it simple and do something like this instead:
```python
import typer


def main(commands: str = typer.Option()):
    cmd_list = commands.split(",")
    print(cmd_list)


if __name__ == "__main__":
    typer.run(main)
```

**Reply by YuriiMotov** [reply, marked as answer]
I don't think it's possible to make it accept values in quotes - shell (?) strips quotation marks and
```
python main.py --command "hello,world","goodbye,world"
```
becomes just 
```
python main.py --command hello,world,goodbye,world
```

You can improve the solution suggested by @heiskane this way (using `callback` instead of `parser`):

```py
from typing import Annotated

import typer

def comma_list(raw: list[str]) -> list[str]:
    return raw[-1].split(",") if (len(raw) > 0) else []

def main(
    command: Annotated[list[str], typer.Option(callback=comma_list)],
):
    print(command)

if __name__ == "__main__":
    typer.run(main)
```

Now `command` argument is properly typed.

**Note:** this implementation takes only last value of `--command`.
So, if you try 
```
python main.py --command hello,world --command goodbye,world
```
it will drop `hello,world` and only take `goodbye,world`.
But you can change `comma_list` to merge all values

**Comment by devraj** [comment]
I implemented something quite similar in my `cli` where it takes multiple values as part of an argument in a standard unix fashion i.e separated by spaces.

Essentially extending on what @heiskane is suggesting you do for `Option` you can define an argument as follows:

```python
from typing import Optional, List
from typing_extensions import Annotated

import typer

@app.command("ack")
async def acknowledge(
    ids: Annotated[List[int], typer.Argument(help="alarm id")], #multiple ids
    message: Annotated[
        Optional[str],
        typer.Option(
            "-m",
            "--message",
            help="comment to add to history",
        ),
    ] = None,
):
    for id in ids:
        print(id)
```

See [issue on my project](https://github.com/anomaly/gallagher/issues/55) for more detail.
